src/trel_dnssd/trel_dnssd.hpp

```diff
@@ -0,0 +1,211 @@
+
+    /**
+     * This method registers the TREL service to DNS-SD.
+     *
+     * @param[in] aPort         The UDP port of TREL service.
+     * @param[in] aTxtData      The TXT data of TREL service.
+     * @param[in] aTxtLength    The TXT length of TREL service.
+     *
+     */
+    void RegisterService(uint16_t aPort, const uint8_t *aTxtData, uint8_t aTxtLength);
+
+    /**
+     * This method removes the TREL service from DNS-SD.
+     *
+     */
+    void UnregisterService(void);
+
+    /**
+     * This method notifies that mDNS Publisher is ready.
+     *
+     */
+    void OnMdnsPublisherReady(void);
+
+private:
+    static constexpr size_t   kPeerCacheSize             = 256;
+    static constexpr uint16_t kCheckNetifReadyIntervalMs = 5000;
+
+    struct RegisterInfo
+    {
+        uint16_t                               mPort = 0;
+        std::vector<Mdns::Publisher::TxtEntry> mTxtEntries;
+        std::string                            mInstanceName;
+
+        bool IsValid(void) const { return mPort > 0; }
+        bool IsRegistered(void) const { return !mInstanceName.empty(); }
+        void Assign(uint16_t aPort, const uint8_t *aTxtData, uint8_t aTxtLength);
+        void Clear(void);
+    };
+
+    using Clock = std::chrono::system_clock;
+
+    struct Peer
+    {
+        static const char kTxtRecordExtAddressKey[];
+
+        explicit Peer(std::vector<uint8_t> aTxtData, const otSockAddr &aSockAddr)
+            : mDiscoverTime(Clock::now())
+            , mTxtData(std::move(aTxtData))
+            , mSockAddr(aSockAddr)
+        {
+            ReadExtAddrFromTxtData();
+        }
+
+        void ReadExtAddrFromTxtData(void);
+
+        Clock::time_point    mDiscoverTime;
+        std::vector<uint8_t> mTxtData;
+        otSockAddr           mSockAddr;
+        otExtAddress         mExtAddr;
+        bool                 mValid = false;
+    };
+
+    using PeerMap = std::map<std::string, Peer>;
+
+    bool        IsInitialized(void) const { return !mTrelNetif.empty(); }
+    bool        IsReady(void) const;
+    void        OnBecomeReady(void);
+    void        CheckTrelNetifReady(void);
+    std::string GetTrelInstanceName(void);
+    void        DoRegisterTrelService(void);
+    void        DoRemoveTrelService(void);
+    void        OnTrelServiceInstanceResolved(const std::string &                            aType,
+                                              const Mdns::Publisher::DiscoveredInstanceInfo &aInstanceInfo);
+    void        OnTrelServiceInstanceAdded(const Mdns::Publisher::DiscoveredInstanceInfo &aInstanceInfo);
+    void        OnTrelServiceInstanceRemoved(const std::string &aInstanceName);
+
+    void     NotifyRemovePeer(const Peer &aPeer);
+    void     CheckPeersNumLimit(void);
+    void     RemoveAllPeers(void);
+    uint16_t CountDuplicatePeers(const Peer &aPeer);
+
+    static TrelDnssd *sTrelDnssd;
+
+    Mdns::Publisher &          mPublisher;
+    Ncp::ControllerOpenThread &mNcp;
```

Comment:
Just curious if it's a common pattern to have references as class members.

Reply:
The same pattern happened in multiple places.  So, I am following the pattern.